pkg/operator/workloads/logs.go

```diff
@@ -133,52 +143,164 @@ func ReadLogs(appName string, workloadID string, verbose bool, socket *websocket
+	cmdPath := "/bin/bash"
+
+	kubectlArgs := []string{"kubectl", "-n=" + config.Cortex.Namespace, "logs", "--follow=true"}
+	if previous {
+		kubectlArgs = append(kubectlArgs, "--previous")
+	}
+
+	identifier := pod.Name
+	kubectlArgs = append(kubectlArgs, pod.Name)
+	if pod.Labels["workloadType"] == WorkloadTypeAPI && pod.Labels["userFacing"] == "true" {
+		kubectlArgs = append(kubectlArgs, apiContainerName)
+		kubectlArgs = append(kubectlArgs, "--since="+podCheckInterval.String())
+		identifier += " " + apiContainerName
+	}
+
+	labelLog := fmt.Sprintf(" | while read -r; do echo \"[%s] $REPLY \" | tail -n +1; done", identifier)
+	kubectlCmd := strings.Join(kubectlArgs, " ")
+	bashArgs := []string{"/bin/bash", "-c", kubectlCmd + labelLog}
+	process, err := os.StartProcess(cmdPath, bashArgs, attrs)
 	if err != nil {
-		errors.Panic(err, strings.Join(args, " "))
+		return nil, errors.Wrap(err, strings.Join(bashArgs, " "))
 	}
 
-	go pumpStdout(socket, outr, verbose, true)
-	pumpStdin(socket, inw)
-	stopProcess(process)
+	return process, nil
+}
+
+func podCheck(podCheckChan chan struct{}, socketWriterError chan error, initialPodList []kcore.Pod, previous bool, inr *os.File, outw *os.File) {
+	defer inr.Close()
+	defer outw.Close()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	processMap := make(map[string]*os.Process)
+	defer deleteProcesses(processMap)
+	labels := initialPodList[0].GetLabels()
+	appName := labels["appName"]
+	workloadID := labels["workloadID"]
+
+	for {
+		select {
+		case <-podCheckChan:
+			return
+		case <-timer.C:
+			pods, err := config.Kubernetes.ListPodsByLabels(map[string]string{
+				"appName":    appName,
+				"workloadID": workloadID,
+				"userFacing": "true",
+			})
+
+			if err != nil {
+				socketWriterError <- errors.Wrap(err, "pod check")
+				return
+			}
+
+			runningPodsMap := make(map[string]kcore.Pod)
+			latestRunningPods := strset.New()
+			for _, pod := range pods {
+				if k8s.GetPodStatus(&pod) != k8s.PodStatusPending {
+					latestRunningPods.Add(pod.GetName())
+					runningPodsMap[pod.GetName()] = pod
+				}
+			}
+
+			runningPods := strset.New()
+
+			for podName := range processMap {
+				runningPods.Add(podName)
+			}
+
+			newPods := strset.Difference(latestRunningPods, runningPods)
+			podsToDelete := strset.Difference(runningPods, latestRunningPods)
+			podsToKeep := strset.Intersection(runningPods, latestRunningPods)
+
+			// Prioritize adding running pods
+			podsToAdd := []string{}
+			podsToAddP2 := []string{}
+			for podName := range newPods {
+				pod := runningPodsMap[podName]
+				if k8s.GetPodStatus(&pod) == k8s.PodStatusRunning {
+					podsToAdd = append(podsToAdd, podName)
+				} else {
+					podsToAddP2 = append(podsToAddP2, podName)
+				}
+			}
+			podsToAdd = append(podsToAdd, podsToAddP2...)
+
+			for _, podName := range podsToAdd {
+				if len(podsToKeep) < maxParallelPodLogging {
```

Comment:
It would be cleaner to do `if len(podsToKeep) >= maxParallelPodLogging {` at the top